File: src/gateway/stockanalysis_connector.py
"""
Connector for fetching financial data from stockanalysis.com.
Scrapes income statement, balance sheet, cash flow, ratios, and overview data.
"""

import logging
import re
from typing import Any

import pandas as pd
import requests
from bs4 import BeautifulSoup

logger = logging.getLogger(__name__)


class StockanalysisConnector:
    """Connector for extracting financial data from stockanalysis.com."""

    BASE_URL = "https://stockanalysis.com/stocks"
    HEADERS = {
        "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36"
    }

    def __init__(self, ticker: str):
        """Initialize connector with a stock ticker.

        Args:
            ticker: Stock ticker symbol (e.g., 'VIST', 'AAPL')
        """
        self.ticker = ticker.upper()
        self.base_url = f"{self.BASE_URL}/{self.ticker.lower()}"

    def _fetch_page(self, url: str) -> BeautifulSoup | None:
        """Fetch and parse a page from stockanalysis.com.

        Args:
            url: Full URL to fetch

        Returns:
            BeautifulSoup object or None if fetch fails
        """
        try:
            response = requests.get(url, headers=self.HEADERS, timeout=10)
            response.raise_for_status()
            return BeautifulSoup(response.content, "html.parser")
        except requests.RequestException as e:
            logger.debug(f"Failed to fetch {url}: {e}")
            return None

    def _parse_number(self, text: str) -> float | str:
        """Parse text to number, handling various formats (1,234.56M, 52.72%, etc)."""
        if not isinstance(text, str):
            return text

        try:
            is_percentage = "%" in text
            text = text.replace("%", "").strip()

            multiplier = 1
            if text.endswith("M"):
                multiplier = 1_000_000
                text = text[:-1]
            elif text.endswith("B"):
                multiplier = 1_000_000_000
                text = text[:-1]
            elif text.endswith("K"):
                multiplier = 1_000
                text = text[:-1]

            text = text.replace(",", "")
            value = float(text) * multiplier

            if is_percentage:
                return value / 100 if value > 1 else value

            return value
        except (ValueError, AttributeError):
            return text

    def _extract_table_data(self, soup: BeautifulSoup) -> pd.DataFrame | None:
        """Extract financial table from HTML."""
        try:
            table = soup.find("table")
            if not table:
                return None

            headers = []
            header_row = table.find("tr")
            if header_row:
                for th in header_row.find_all(["th", "td"]):
                    headers.append(th.get_text(strip=True))

            if not headers:
                return None

            rows = []
            for tr in table.find_all("tr")[1:]:
                row_data = []
                for td in tr.find_all("td"):
                    text = td.get_text(strip=True)
                    text = self._parse_number(text)
                    row_data.append(text)
                if row_data and len(row_data) == len(headers):
                    rows.append(row_data)

            if not rows:
                return None

            return pd.DataFrame(rows, columns=headers)

        except Exception as e:
            logger.debug(f"Failed to extract table data: {e}")
            return None

    def get_overview(self) -> dict[str, Any]:
        """Get company overview and key metrics."""
        url = f"{self.base_url}/"
        soup = self._fetch_page(url)

        if not soup:
            return {"ticker": self.ticker}

        overview_data = {"ticker": self.ticker, "url": url}

        try:
            company_name = soup.find("h1")
            if company_name:
                overview_data["name"] = company_name.get_text(strip=True)

            page_text = soup.get_text()

            patterns = {
                "price": r"Price[:\s]+\$?([\d,]+\.?\d*)",
                "market_cap": r"Market Cap[:\s]+\$?([\d,]+[MBK]?)",
                "pe_ratio": r"P\/E Ratio[:\s]+([\d,]+\.?\d*)",
            }

            for key, pattern in patterns.items():
                match = re.search(pattern, page_text, re.IGNORECASE)
                if match:
                    overview_data[key] = self._parse_number(match.group(1))

        except Exception as e:
            logger.debug(f"Error extracting overview data: {e}")

        return overview_data

    def get_income_statement(self, period: str = "quarterly") -> pd.DataFrame | None:
        """Get income statement data."""
        period_param = "yearly" if period == "annual" else "quarterly"
        url = f"{self.base_url}/financials/?p={period_param}"

        soup = self._fetch_page(url)
        if not soup:
            return None

        df = self._extract_table_data(soup)
        return df.set_index(df.columns[0]) if (df is not None and len(df) > 0) else None

    def get_balance_sheet(self, period: str = "quarterly") -> pd.DataFrame | None:
        """Get balance sheet data."""
        period_param = "yearly" if period == "annual" else "quarterly"
        url = f"{self.base_url}/financials/balance-sheet/?p={period_param}"

        soup = self._fetch_page(url)
        if not soup:
            return None

        df = self._extract_table_data(soup)
        return df.set_index(df.columns[0]) if (df is not None and len(df) > 0) else None

    def get_cash_flow_statement(self, period: str = "quarterly") -> pd.DataFrame | None:
        """Get cash flow statement data."""
        period_param = "yearly" if period == "annual" else "quarterly"
        url = f"{self.base_url}/financials/cash-flow-statement/?p={period_param}"

        soup = self._fetch_page(url)
        if not soup:
            return None

        df = self._extract_table_data(soup)
        return df.set_index(df.columns[0]) if (df is not None and len(df) > 0) else None

    def get_ratios(self, period: str = "quarterly") -> pd.DataFrame | None:
        """Get financial ratios data."""
        period_param = "yearly" if period == "annual" else "quarterly"
        url = f"{self.base_url}/financials/ratios/?p={period_param}"

        soup = self._fetch_page(url)
        if not soup:
            return None

        df = self._extract_table_data(soup)
        return df.set_index(df.columns[0]) if (df is not None and len(df) > 0) else None

    def get_all_data(self, period: str = "quarterly") -> dict[str, Any]:
        """Get all financial data in one call."""
        return {
            "overview": self.get_overview(),
            "income_statement": self.get_income_statement(period),
            "balance_sheet": self.get_balance_sheet(period),
            "cash_flow": self.get_cash_flow_statement(period),
            "ratios": self.get_ratios(period),
        }

File: src/gateway/stockanalysisConnector.ts
import * as cheerio from "cheerio";

import { DataFetchError, DataSource } from "../domain";
import { HttpClient } from "./httpClient";
import {
  ApiGateway,
  FetchStockDataInput,
  FetchStockDataOutput,
  FinancialStatementData,
  StockDataFromSource,
} from "./types";

export class StockanalysisConnector implements ApiGateway {
  readonly name = "stockanalysis";
  private readonly httpClient: HttpClient;
  private isHealthyFlag = true;
  private lastErrorValue?: Error;
  private lastErrorTimeValue?: Date;

  constructor() {
    this.httpClient = new HttpClient({
      baseURL: "https://stockanalysis.com",
      timeout: 20000,
      retryAttempts: 2,
    });
  }

  get isHealthy(): boolean {
    return this.isHealthyFlag;
  }

  get lastError(): Error | undefined {
    return this.lastErrorValue;
  }

  get lastErrorTime(): Date | undefined {
    return this.lastErrorTimeValue;
  }

  markHealthy(): void {
    this.isHealthyFlag = true;
    this.lastErrorValue = undefined;
    this.lastErrorTimeValue = undefined;
  }

  markUnhealthy(error: Error): void {
    this.isHealthyFlag = false;
    this.lastErrorValue = error;
    this.lastErrorTimeValue = new Date();
  }

  async fetchStockData(input: FetchStockDataInput): Promise<FetchStockDataOutput> {
    const { ticker, period } = input;

    try {
      const url = `/stocks/${ticker}/${period}/`;
      const response = await this.httpClient.get<string>(url, DataSource.StockAnalysis);

      const stockData = this.parseHtmlResponse(response.data as unknown as string, ticker);
      this.markHealthy();

      return {
        data: stockData,
        source: DataSource.StockAnalysis,
        fetchedAt: new Date(),
      };
    } catch (error) {
      const err = error instanceof Error ? error : new Error(String(error));
      this.markUnhealthy(err);
      throw new DataFetchError(
        DataSource.StockAnalysis,
        `Failed to fetch stock data for ${ticker}`,
        { ticker, period: input.period, originalError: err.message },
      );
    }
  }

  private parseHtmlResponse(html: string, ticker: string): StockDataFromSource {
    const $ = cheerio.load(html);

    const marketCapText = this.extractText($, '[data-key="market-cap"]');
    const peRatioText = this.extractText($, '[data-key="pe-ratio"]');
    const epsText = this.extractText($, '[data-key="eps"]');
    const dividendYieldText = this.extractText($, '[data-key="dividend-yield"]');

    return {
      overview: {
        ticker,
        name: this.extractCompanyName($),
        sector: this.extractText($, '[data-key="sector"]'),
        marketCap: marketCapText ? this.parseNumber(marketCapText) : undefined,
        peRatio: peRatioText ? this.parseNumber(peRatioText) : undefined,
        eps: epsText ? this.parseNumber(epsText) : undefined,
        dividendYield: dividendYieldText ? this.parseNumber(dividendYieldText) : undefined,
      },
      incomeStatement: this.extractFinancialTable($, "income-statement"),
      balanceSheet: this.extractFinancialTable($, "balance-sheet"),
      cashFlow: this.extractFinancialTable($, "cash-flow"),
      ratios: this.extractFinancialTable($, "ratios"),
      statistics: this.extractFinancialTable($, "statistics"),
      peers: this.extractPeers($),
    };
  }

  private extractCompanyName($: cheerio.CheerioAPI): string {
    return $("h1").text().trim().split(" ")[0] || "";
  }

  private extractText($: cheerio.CheerioAPI, selector: string): string | undefined {
    const text = $(selector).text().trim();
    return text || undefined;
  }

  private parseNumber(value: string | undefined): number | undefined {
    if (!value) return undefined;

    const multipliers: Record<string, number> = {
      T: 1e12,
      B: 1e9,
      M: 1e6,
      K: 1e3,
    };

    const match = value.match(/^([\d.]+)([TBMK])?/);
    if (!match || !match[1]) return undefined;

    const num = Number.parseFloat(match[1]);
    const multiplier = match[2] ? (multipliers[match[2]] ?? 1) : 1;

    return num * multiplier;
  }

  private extractFinancialTable(
    $: cheerio.CheerioAPI,
    tableId: string,
  ): readonly FinancialStatementData[] {
    const rows: FinancialStatementData[] = [];
    const table = $(`[data-table="${tableId}"] table`);

    table.find("tbody tr").each((_, row) => {
      const cells = $(row).find("td");
      if (cells.length > 0) {
        const rowData: Record<string, string | undefined> = {};
        cells.each((index, cell) => {
          const key = `col${index}`;
          const text = $(cell).text().trim();
          rowData[key] = text || undefined;
        });
        rows.push(rowData as FinancialStatementData);
      }
    });

    return rows;
  }

  private extractPeers($: cheerio.CheerioAPI): string[] | undefined {
    const peersSection = $('[data-section="peers"]');
    if (peersSection.length === 0) return undefined;

    const peers: string[] = [];
    peersSection.find("a[href*='/stocks/']").each((_, link) => {
      const ticker = $(link).text().trim();
      if (ticker) peers.push(ticker);
    });

    return peers.length > 0 ? peers : undefined;
  }
}
